src/artifact_providers/github.ts

```diff
@@ -110,25 +109,51 @@ export class GitHubArtifactProvider extends BaseArtifactProvider {
         lastArtifact.created_at >= revisionDate;
     }
 
-    if (checkNextPage) {
-      page += 1;
-    } else {
-      // If we are retrying, reset page index and start over
-      page = 0;
-      tries += 1;
-    }
-
-    if (tries < MAX_TRIES) {
-      return this.getRevisionArtifact(revision, revisionDate, page, tries);
-    }
+    return null;
+  }
 
-    if (artifactResponse.total_count === 0) {
-      throw new Error(`Failed to discover any artifacts (tries: ${tries})`);
-    } else {
-      throw new Error(
-        `Can't find any artifacts for revision "${revision}" (tries: ${tries})`
+  /**
+   * Tries to find the artifact with the given revision, retrying if
+   * necessary.
+   *
+   * @param revision
+   * @returns The artifact for the given revision or throws an error
+   */
+  protected async getRevisionArtifact(
+    revision: string
+  ): Promise<ArtifactItem> {
+    const { repoName: repo, repoOwner: owner } = this.config;
+    let artifact;
+    const getRevisionDate = lazyRequest<string>(async () => {
+      return (await this.github.git.getCommit({
+        owner,
+        repo,
+        commit_sha: revision,
+      })).data.committer.date;
+    })
+
+    for (let tries = 0; tries < MAX_TRIES; tries++) {
+      this.logger.info(
+        `Fetching GitHub artifacts for ${owner}/${repo}, revision ${revision} (attempt ${tries + 1} of ${MAX_TRIES})`
       );
+
+      artifact = await this.searchForRevisionArtifact(revision, getRevisionDate);
+      if (artifact) {
+        return artifact;
+      }
+
+      // There may be a race condition between artifacts being uploaded
+      // and the GitHub API having the info to return.
+      // Wait before retries to give GitHub a chance to propagate changes.
+      if (tries + 1 < MAX_TRIES) {
```

Comment:
Also, now that I'm thinking more clearly, I don't think there's any point going through all the pages in retries:

If we assume the response is order by descending date and we couldn't find the artifact in our first pagination, it is impossible for it to show up on any page except for the first page, right?

Reply:
in a repository with lots of activity and lots of artifacts it may appear paginated (since this api is the entire repository)

Reply:
For that to happen you need 100+ artifacts showing up in a matter of seconds. Don't think that's very realistic but yeah, valid point.